src/utils/embedded/embedded.utils.ts

```diff
@@ -76,15 +82,127 @@ const supabase = IS_EMBEDDED_APP
+async function getSessionId() {
+  try {
+    const {
+      data: { session }
+    } = await supabase.auth.getSession();
+    const jwtDecoded = jwtDecode(session?.access_token) as {
+      session_id: string;
+    };
+    return jwtDecoded.session_id;
+  } catch (err) {
+    return undefined;
+  }
+}
+
+async function insecurelyValidateApplication() {
+  try {
+    const sessionId = await getSessionId();
+    const result = await trpcVanilla.validateApplication.query({
+      clientId: EMBEDDED_CLIENT_ID,
+      applicationId: EMBEDDED_APPLICATION_ID,
+      applicationOrigin: EMBEDDED_ANCESTOR_ORIGIN,
+      sessionId
+    });
+
+    return result;
+  } catch (err: any) {
+    // Only show errors if we're inside an iframe.
+    // TODO:Should we show different errors or instructions if we are not inside an iframe?
+    if (!isInsideIframe()) return;
+
+    // Check if error message already exists
+    const existingError = document.querySelector("[data-wander-error]");
+    if (existingError) {
+      return;
+    }
+
+    // Only show errors for validation failures
+    // TRPC errors will have data.code property
+    if (
+      !err.data?.code ||
+      !["NOT_FOUND", "BAD_REQUEST", "FORBIDDEN"].includes(err.data.code)
+    ) {
+      console.error("Unexpected error during validation:", err);
+      return;
+    }
+
+    const errorMessages = {
+      NOT_FOUND:
+        "Invalid application configuration. Please verify your applicationId and clientId.",
+      BAD_REQUEST: `Invalid origin URL provided`,
+      FORBIDDEN:
+        err.message || "This domain is not authorized to use this application."
+    };
+
+    const errorMessage = errorMessages[err.data.code];
+    if (!errorMessage) {
+      return;
+    }
+
+    const overlay = document.createElement("div");
+    overlay.setAttribute("data-wander-error", "overlay");
+    overlay.style.cssText = `
+      position: fixed;
+      inset: 0;
+      background: rgba(0, 0, 0, 0.5);
+      backdrop-filter: blur(2px);
+      z-index: 99998;
+    `;
+
+    const errorPopover = document.createElement("div");
+    errorPopover.setAttribute("data-wander-error", "popover");
+    errorPopover.style.cssText = `
+      position: fixed;
+      top: 50%;
+      left: 50%;
+      transform: translate(-50%, -50%);
+      background: #1a1a1a;
+      color: white;
+      font-family: system-ui;
+      text-align: center;
+      padding: 16px 24px;
+      border-radius: 8px;
+      box-shadow: 0 4px 12px rgba(0,0,0,0.3);
+      z-index: 99999;
+      max-width: 400px;
+      width: calc(100% - 32px);
+    `;
+
+    errorPopover.innerHTML = `
+      <h1 style="
+        color: #ff5757;
+        margin: 0 0 8px 0;
+        font-size: 16px;
+        line-height: 1.4;
+      ">Invalid Wander Embedded configuration</h1>
+      <p style="
+        margin: 0;
+        font-size: 14px;
+        line-height: 1.5;
+        opacity: 0.9;
+      ">${errorMessage}</p>
+    `;
+
+    document.body.appendChild(overlay);
```

Comment:
Rather than appending it, I think it's better to replace the whole document, that's why I was using `document.write`. The idea is that this looks as close as possible to what users would get it validation was happening on the server side, meaning, if it fails, the app won't even load.